dali/kernels/common/pad.h

```diff
@@ -0,0 +1,146 @@
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+#include "dali/kernels/slice/slice_flip_normalize_permute_common.h"
+#include "dali/kernels/slice/slice_flip_normalize_permute_pad_cuda_impl.cuh"
+
+namespace dali {
+namespace kernels {
+
+template <typename Type, size_t Dims>
+class DLL_PUBLIC PadGPU {
+ private:
+  static constexpr size_t kBlockDim = 512;
+  static constexpr size_t kBlockSize = 64 * kBlockDim;
+  size_t block_count_ = 0;
+
+ public:
+  DLL_PUBLIC PadGPU() = default;
+
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<Type, Dims> &in,
+                                      int axis) {
+    KernelRequirements req;
+    if (in.num_elements() == 0) {
+      req.output_shapes = {in.shape};
+    } else {
+      auto biggest_dim = in.tensor_shape(0)[axis];
+      auto num_samples = static_cast<size_t>(in.num_samples());
+      for (size_t i = 1; i < num_samples; ++i) {
+        auto dim = in.tensor_shape(i)[axis];
+        if (dim > biggest_dim)
+          biggest_dim = dim;
+      }
+      auto ts = in.tensor_shape(0);
+      ts[axis] = biggest_dim;
+
+      std::vector<TensorShape<DynamicDimensions>> tl_shape(num_samples, ts);
+      req.output_shapes = {TensorListShape<DynamicDimensions>(tl_shape)};
+      block_count_ = 0;
+      for (size_t i = 0; i < num_samples; ++i) {
+        block_count_ += std::ceil((volume(ts)) / static_cast<float>(kBlockSize));
+      }
+      ScratchpadEstimator se;
+      se.add<detail::SampleDesc<Dims>>(AllocType::Host, num_samples);
+      se.add<detail::SampleDesc<Dims>>(AllocType::GPU, num_samples);
+
+      se.add<detail::BlockDesc>(AllocType::Host, block_count_);
+      se.add<detail::BlockDesc>(AllocType::GPU, block_count_);
+      req.scratch_sizes = se.sizes;
+    }
+    return req;
+  }
+
+  DLL_PUBLIC void Run(KernelContext &context, OutListGPU<Type, Dims> &out,
+                      const InListGPU<Type, Dims> &in, Type padding_val) {
+    auto num_samples = static_cast<size_t>(in.num_samples());
+
+    detail::SampleDesc<Dims>* sample_descs_cpu =
+      context.scratchpad->Allocate<detail::SampleDesc<Dims>>(AllocType::Host,
+                                                             num_samples);
+    detail::BlockDesc *block_descs_cpu =
+      context.scratchpad->Allocate<detail::BlockDesc>(AllocType::Host, block_count_);
+
+    std::vector<size_t> sample_sizes(in.size());
+    for (int i = 0; i < in.size(); i++) {
+      const auto in_shape = in.tensor_shape(i);
+      auto out_shape = out.tensor_shape(i);
+
+      auto &sample_desc = sample_descs_cpu[i];
+      sample_desc.in_strides = GetStrides<Dims>(in_shape);
+      sample_desc.out_strides = GetStrides<Dims>(out_shape);
+      // Weird implem: we rely only on strides but we need out_shape != padded_out_shape
```

Comment:
I don't understand this comment.
If padded out shape is bigger than out shape, then we pad, otherwise no padding takes place

Reply:
Implem temporary comment I forgot to remove :sweat_smile: